26_Autoexposure/main.cpp

```diff
@@ -0,0 +1,874 @@
+					.dstStageMask = PIPELINE_STAGE_FLAGS::COPY_BIT,
+					.dstAccessMask = ACCESS_FLAGS::TRANSFER_WRITE_BIT
+				},
+				{
+					.srcStageMask = PIPELINE_STAGE_FLAGS::COPY_BIT,
+					.srcAccessMask = ACCESS_FLAGS::TRANSFER_WRITE_BIT,
+				}
+			};
+
+			// upload image and write to descriptor set
+			queue->startCapture();
+
+			cmdbuf->begin(IGPUCommandBuffer::USAGE::ONE_TIME_SUBMIT_BIT);
+			// change the layout of the image
+			const IGPUCommandBuffer::SImageMemoryBarrier<IGPUCommandBuffer::SOwnershipTransferBarrier> imgBarriers1[] = {
+				{
+					.barrier = {
+						.dep = transferBarriers[0]
+						// no ownership transfers
+					},
+					.image = m_gpuImg.get(),
+					// transition the whole view
+					.subresourceRange = cpuImgParams.subresourceRange,
+					// a wiping transition
+					.newLayout = IGPUImage::LAYOUT::TRANSFER_DST_OPTIMAL
+				}
+			};
+			const IGPUCommandBuffer::SImageMemoryBarrier<IGPUCommandBuffer::SOwnershipTransferBarrier> imgBarriers2[] = {
+				{
+					.barrier = {
+						.dep = transferBarriers[1]
+						// no ownership transfers
+					},
+					.image = m_gpuImg.get(),
+					// transition the whole view
+					.subresourceRange = cpuImgParams.subresourceRange,
+					// a wiping transition
+					.oldLayout = IGPUImage::LAYOUT::TRANSFER_DST_OPTIMAL,
+					.newLayout = IGPUImage::LAYOUT::READ_ONLY_OPTIMAL
+				}
+			};
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, { .imgBarriers = imgBarriers1 });
+			// upload contents
+			m_utils->updateImageViaStagingBuffer(
+				m_intendedSubmit,
+				cpuImgParams.image->getBuffer(),
+				cpuImgParams.image->getCreationParameters().format,
+				m_gpuImg.get(),
+				IGPUImage::LAYOUT::TRANSFER_DST_OPTIMAL,
+				cpuImgParams.image->getRegions()
+			);
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, { .imgBarriers = imgBarriers2 });
+			m_utils->autoSubmit(m_intendedSubmit, [&](SIntendedSubmitInfo& nextSubmit) -> bool { return true; });
+
+			IGPUImageView::SCreationParams gpuImgViewParams = {
+				.image = m_gpuImg,
+				.viewType = IGPUImageView::ET_2D,
+				.format = m_gpuImg->getCreationParameters().format
+			};
+
+			m_gpuImgView = m_device->createImageView(std::move(gpuImgViewParams));
+
+			IGPUDescriptorSet::SDescriptorInfo info1 = {};
+			info1.info.image.imageLayout = IImage::LAYOUT::READ_ONLY_OPTIMAL;
+			info1.desc = m_gpuImgView;
+
+			IGPUDescriptorSet::SDescriptorInfo info2 = {};
+			info2.info.image.imageLayout = IImage::LAYOUT::READ_ONLY_OPTIMAL;
+			info2.desc = m_gpuImgView; // FIXME: temporarily pass in input image
+
+			IGPUDescriptorSet::SWriteDescriptorSet writeDescriptors[] = {
+				{
+					.dstSet = m_lumaPresentDS[0].get(),
+					.binding = 0,
+					.arrayElement = 0,
+					.count = 1,
+					.info = &info1
+				},
+				{
+					.dstSet = m_lumaPresentDS[1].get(),
+					.binding = 0,
+					.arrayElement = 0,
+					.count = 1,
+					.info = &info2
+				}
+			};
+
+			m_device->updateDescriptorSets(2, writeDescriptors, 0, nullptr);
```

Comment:
see https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/137/files#r1719794211